src/types.rs

```diff
@@ -347,12 +349,68 @@ pub enum PlayerClass {
     Void,
 }
 
+#[derive(Debug, Clone)]
+pub struct Position {
+    x: f32,
+    y: f32,
+    z: f32,
+}
+
+impl<'de> Deserialize<'de> for Position {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct PositonVisitor {}
+
+        impl<'de> Visitor<'de> for PositonVisitor {
+            type Value = Position;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("a 3 element list of f32 seperated by space")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let mut coords = [0f32; 3];
+                let splits = s.split(" ").collect::<Vec<&str>>();
+                if splits.len() != 3 {
+                    return Err(serde::de::Error::invalid_length(splits.len(), &self));
+                }
+
+                for (i, s) in splits.iter().enumerate() {
+                    coords[i] = s.parse().map_err(|_| serde::de::Error::invalid_type(serde::de::Unexpected::Str(s), &self))?;
+                }
+                let [x, y, z] = coords;
+                Ok(Position { x, y, z })
```

Comment:
```suggestion
                let splits = s.split(" ").collect::<Vec<&str>>();
                if splits.len() != 3 {
                    return Err(serde::de::Error::invalid_length(splits.len(), &self));
                }

                let parse = |split: &str| split.parse().map_err(|_| serde::de::Error::invalid_type(serde::de::Unexpected::Str(s), &self));
                Ok(Position {
                    x: parse(splits[0])?,
                    y: parse(splits[1])?,
                    z: parse(splits[2])?,
                })
```
This is a bit cleaner imo, since we don't use an intermediate array.